packages/teleterm/src/ui/utils/retryWithRelogin.ts

```diff
@@ -0,0 +1,71 @@
+import {} from 'teleterm/ui/services/workspacesService/documentsService';
+import { routing } from 'teleterm/ui/uri';
+import AppContext from 'teleterm/ui/appContext';
+
+export async function retryWithRelogin<T>(
+  appContext: AppContext,
+  originatingDocumentUri: string,
+  resourceUri: string,
+  actionToRepeat: () => Promise<T>
+): Promise<T> {
+  let retryableErrorFromActionToRepeat: Error;
+  try {
+    return await actionToRepeat();
+  } catch (error) {
+    // TODO(ravicious): Replace this with actual check on metadata.
+    const isRetryable =
+      error instanceof Error &&
+      (error.message.includes('ssh: handshake failed') ||
+        error.message.includes('ssh: cert has expired'));
+
+    if (isRetryable) {
+      retryableErrorFromActionToRepeat = error;
+    } else {
+      throw error;
+    }
+  }
+
+  const isDocumentStillActive = appContext.workspacesService.isDocumentActive(
+    originatingDocumentUri
+  );
+
+  if (!isDocumentStillActive) {
+    // The error is retryable, but the user is no longer looking at the relevant UI, for example
+    // they switched to a different document or a different workspace completely.
+    //
+    // Since it might take a few seconds before the execution gets to this point, the user might no
+    // longer remember what their intent was, thus showing them a login modal can be disorienting.
+    //
+    // In that situation, let's just not attempt to relogin and instead let's surface the original
+    // error.
+    throw retryableErrorFromActionToRepeat;
+  }
+
+  const rootClusterUri = routing.ensureRootClusterUri(resourceUri);
+
+  try {
+    await login(appContext, rootClusterUri);
+  } catch (error) {
+    // In case the user canceled the login attempt, retry the original call anyway in case the cert
+    // was refreshed externally, for example through tsh.
+    const isRetryable = error instanceof LoginAttemptCanceledError;
+
+    if (!isRetryable) {
+      throw error;
+    }
+  }
```

Comment:
While writing tests I realized that the custom error class is completely unnecessary. The login attempt either succeeds or the user closes it, so the promise can be just resolved/rejected.

We don't see any errors related to the process of logging in, because those are contained within the login modal.

I fully control when and why the promise is rejected.

Reply:
Well, technically `appContext.modalsService.openClusterConnectDialog` could throw an error and thus reject the promise with something else than `LoginAttemptCanceledError`. But we're going to assume that such behavior is a bug in the program rather than something we need to handle.

Reply:
To avoid using exceptions for control flow, we could just resolve the promise to `true` when the user logs in and `false` if they cancel the attempt. Then if the promise is rejected, that means there was an unexpected error somewhere else.

Reply:
Matter of fact, I don't actually care in this case if they canceled the attempt or not, because I'm going to repeat the call anyway. So I'll just resolve to void in both cases and any unexpected errors will bubble up thanks to `await`.

Reply:
Returning `true`/`false` sounds good to me.

Why do we want to repeat do call when user cancels the attempt?

Reply:
https://github.com/gravitational/webapps/blob/6ef6affc0d23065ac23a2f22f29733fdf11dc347/packages/teleterm/src/ui/utils/retryWithRelogin.ts#L46-L50

I actually removed that comment by mistake, let me bring it back. ;)